### PR TITLE
Reprocess Step Function StartExecution call throttling

### PR DESCRIPTION
**Pull request by amork**
Connects to #1062
TODOs:

- [x] script dead_letter_queue_target_arn in config.json
- [x] exit condition, reprocess count

### Test plan
Ran series of scale test. Identified issue with describeExecution throttling in the checkout service #1094 There are still failures that are being investigated. More testing after #1094 is implemented

### CODE REVIEW

File: dss/stepfunctions/__init__.py [hunk 1]
```diff
@@ -1,9 +1,29 @@
...
 import os
 import typing
+import logging
 
 from dss.util.aws import ARN
 from dss.util.aws.clients import stepfunctions  # type: ignore
+from dss.util.aws import send_sns_msg
+
+"""
+The keys used to transfer step function invokation data over SNS to the dss-sfn-* Lambda. dss-sfn starts step function
```
Comment (xbrianh) [inline dss/stepfunctions/__init__.py]: invokation -> invocation

File: daemons/dss-sfn/app.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+
+@app.scheduled_function("rate(1 minute)")
+def reaper(event, context):
+    queue_name = "dss-dlq-sfn-" + os.environ["DSS_DEPLOYMENT_STAGE"]
+
+    # Get the queue
+    queue = sqs.get_queue_by_name(QueueName=queue_name)
+
+    for message in queue.receive_messages():
```
Comment (xbrianh) [inline daemons/dss-sfn/app.py]: This will only process 1 message per reaper invocation unless `MaxNumberOfMessages` is set higher?

Comment (amork) [reply daemons/dss-sfn/app.py]: fixed, thanks

File: tests/scalability/tester.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (amork) [reply tests/scalability/tester.py]: I did not mean to include this file into PR. Deleting

File: dss/stepfunctions/__init__.py (hunk 1 above)
Comment (amork) [reply dss/stepfunctions/__init__.py]: fixed

File: daemons/dss-dlq-reaper/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# HCA DSS: Dead-letter queue based retry framework
+
+This daemon is a part of the DLQ-based framework for reprocessing failed Lambda invokations
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-dlq-reaper/README.md]: fixed

File: daemons/dss-dlq-reaper/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+#### About the DLQ retry framework
+
+1. The retry framework is based on the DLQ support built into AWS Lambda. When Lambda invocation fails the message that
+triggered the invocation is placed into DLQ
+2. The daemon is triggred by cloudwatch cron every minute
+3. The daemon retrieves messages from SQS queue  dss-dlq-{stage}
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-dlq-reaper/README.md]: fixed

File: daemons/dss-dlq-reaper/app.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+def reaper(event, context):
+    queue_name = "dss-dlq-" + os.environ["DSS_DEPLOYMENT_STAGE"]
+
+    # Get the queue
+    queue = sqs.get_queue_by_name(QueueName=queue_name)
+
+    message_count = 0
+
+    for i in range(RECEIVE_BATCH_COUNT):
+        for message in queue.receive_messages(MaxNumberOfMessages=10, AttributeNames=['All'],
```
[the remark this answers is not in the document]
Comment (amork) [reply daemons/dss-dlq-reaper/app.py]: Valid values are 1 to 10. Default is 1.

File: dss/stepfunctions/__init__.py [hunk 7]
```diff
@@ -26,21 +43,36 @@ def step_functions_execution_arn(state_machine_name_template: str, execution_nam
...
+        SFN_INPUT_KEY: json.dumps(input)
+    }
+    logger.debug('Sending message: ' + str(message))
+
+    response = send_sns_msg(sfn_sns_topic_arn, message, attributes)
+
+    return response
+
+
+def _step_functions_start_execution(state_machine_name_template: str, execution_name: str,
```
[the remark this answers is not in the document]
Comment (amork) [reply dss/stepfunctions/__init__.py]: I used _ to indicate that this should not be used in the client code. This function is invoked by dss-sfn-launcher Lambda only.
There is a function used by the client code with same name without _.
Do you have any recommendations on the alternative naming?

File: dss/stepfunctions/__init__.py [hunk 8]
```diff
@@ -62,9 +94,9 @@ def step_functions_describe_execution(state_machine_name_template: str, executio
 
 
 def step_functions_list_executions(
-    state_machine_name_template: str,
-    status_filter: str=None,
-    max_results_per_page: int=None
+        state_machine_name_template: str,
+        status_filter: str = None,
+        max_results_per_page: int = None
```
[the remark this answers is not in the document]
Comment (amork) [reply dss/stepfunctions/__init__.py]: PyCharm strikes again. Reverted.

File: daemons/dss-dlq-reaper/app.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+@app.scheduled_function("rate(1 minute)")
+def reaper(event, context):
+    queue_name = "dss-dlq-" + os.environ["DSS_DEPLOYMENT_STAGE"]
+
+    # Get the queue
+    queue = sqs.get_queue_by_name(QueueName=queue_name)
+
+    message_count = 0
+
+    while context.get_remaining_time_in_millis() > 10000:
```
Comment (hannes-ucsc) [inline daemons/dss-dlq-reaper/app.py]: You should account for the time it takes to long poll *and* to process each of the max. of ten messages. IOW, leave a little bit of head room.

Comment (kislyuk) [reply daemons/dss-dlq-reaper/app.py]: Agreed, maybe make this 20000.

File: dss/stepfunctions/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kislyuk) [reply dss/stepfunctions/__init__.py]: Then consider moving this function inside the lambda's app.py.

File: scripts/deploy_dss_sfn_queue.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+import boto3
+import os
+
+sqs = boto3.resource('sqs')
+
+# Create the queue. This returns an SQS.Queue instance.
+# Amazon SQS returns error only if the request includes attributes whose values differ from those of the existing queue.
+queue = sqs.create_queue(QueueName='dss-dlq-' + os.environ["DSS_DEPLOYMENT_STAGE"], Attributes={'DelaySeconds': '5'})
+
+print(f"Queue: {queue.url}")
```
Comment (kislyuk) [inline scripts/deploy_dss_sfn_queue.py]: Please replace this script with an AWS CLI invocation. It's a one liner that doesn't warrant a separate script.

Any script in scripts/ should not do anything when run with no arguments, and should print a help message when run with `--help`. (I know this is not true for some scripts in there right now - that is something we should fix)

**Pull request merged**